#72 Gitlab integration

Marios85 opened the pull request:
PR to close #72 thereby adding Gitlab integration to Bodywork

Added support for Gitlab by enabling dynamic addition of host keys to the SSH known_hosts file. 

Before adding a new Git provider to the list of trusted hosts, their fingerprint is checked and cross referenced against a hard-coded copy of their fingerprint (publicly available on their websites) to ensure that the key comes from a genuine source.

Tests have been updated to work by reading in the SSH private key file from a user into the private key environment variable.

Also included in this PR:

  - Integration Test refactoring.
  - Amending the CircleCI config so that the test results are stored in a readable format.
  - Reformatting the whole code base with Black.

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +50,160 @@ def download_project_code_from_repo(
+                encoding="utf-8",
+            ).stdout
+            fingerprint = run(
+                ["ssh-keygen", "-l", "-f", "-"],
```
What's does the `"-"` do?

I couldn't work it out from reading the ssh-keygen man pages.

Marios85 replied:
It won't be in there, it's required for piping in the input from the previous command using the input=server_key.

AlexIoannides replied:
Is that a generic Linux thing?

Is there any chance of making it more explicit, because I'll  never remember that.

Marios85 replied:
I think it is and it works on Windows as well. I don't really know how to make it more explicit apart from in the comments. Would an explanation in the comment suffice?

AlexIoannides replied:
A comment at the very least please.

I thought you could just pass is the filename from the previous command that it's implicitly referring to. Have I misunderstood what's going on?

Marios85 replied:
That would involve creating a file from the previous command and I wanted to avoid creating a file just for the sake of passing it into the keygen command. I think this is the more robust solution as it removes any file handling.

AlexIoannides commented on tests/unit_and_functional/conftest.py:
```diff
@@ -61,11 +72,12 @@ def setup_bodywork_test_project(
     finally:
         # TEARDOWN
         shutil.rmtree('{}/.git'.format(project_repo_location), onerror=on_error)
+        shutil.rmtree('{}/.git'.format(cloned_project_repo_location), ignore_errors=True, onerror=on_error)
```
Could we stick to using f-strings - i.e. `f'{cloned_project_repo_location}/.git'`. I believe that the `.format` approach is slowly being phased-out. I failed to notice this in the above line when I reviewed it the first time around.

Marios85 replied:
Yeah this shouldn't be different from the rest.

AlexIoannides commented on VERSION:
```diff
@@ -1 +1 @@
-2.0.2
+2.1.0
```
2.1.0alpha1

Marios85 replied:
Oops 😬

AlexIoannides commented on tests/unit_and_functional/conftest.py:
```diff
@@ -81,6 +93,26 @@ def k8s_env_vars() -> Iterable[bool]:
+@fixture(scope='function')
+def set_gitlab_ssh_private_key_env_var() -> None:
+    if SSH_PRIVATE_KEY_ENV_VAR not in os.environ:
+        private_key = Path.home() / '.ssh/id_rsa_e28827a593edd69f1a58cf07a7755107'
```
Where does this funny name come from `id_rsa_e28827a593edd69f1a58cf07a7755107`?

Marios85 replied:
That is the filename that CircleCI gives the key (a hash of some sorts) because we now have multiple RSA keys. We have no control over this.

AlexIoannides replied:
Won't the tests fail locally then?

AlexIoannides commented on tests/unit_and_functional/test_git.py:
```diff
@@ -24,22 +24,22 @@
     setup_bodywork_test_project: Iterable[bool],
     project_repo_connection_string: str,
-    cloned_project_repo_location: Path
+    cloned_project_repo_location: Path,
```
I'm not sure that comma is meant to be there.

Marios85 replied:
Good spot!

Marios85 replied:
It's Black that's adding the comma's...I can't say I agree with that.

AlexIoannides commented on tests/unit_and_functional/test_git.py:
```diff
@@ -48,74 +48,113 @@ def test_that_git_project_repo_can_be_cloned(
+        ssh_dir = Path(".") / SSH_DIR_NAME
+        os.makedirs(ssh_dir, exist_ok=False)
+        with open(os.path.join(ssh_dir, "id_rsa"), "w"):
+            pass
```
`pass`?

Is this a WIP?

Marios85 replied:
This is no longer needed, it was required to create a blank file in my previous implementation when the env variable for the private key was removed. 

Another good spot! 🙂

AlexIoannides commented on .circleci/config.yml:
```diff
@@ -28,36 +28,50 @@ jobs:
           command: | 
             git config --global user.name "Bodywork Machine Learning Ltd"
             git config --global user.email "info@bodyworkml.com"
+      - add_ssh_keys:
```
We don't need this now the tests are split?

Marios85 replied:
Yes, this PR isn't ready to review yet so I hadn't cleaned up the config.yml yet.

AlexIoannides commented on .circleci/config.yml:
```diff
@@ -28,36 +28,50 @@ jobs:
     machine: true
     steps:
       - checkout
+      - add_ssh_keys:
```
Should no longer be needed.

AlexIoannides commented on .circleci/config.yml:
```diff
@@ -68,6 +82,10 @@ jobs:
     machine: true
     steps:
       - checkout
+      - add_ssh_keys:
```
Delete.

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +50,162 @@ def download_project_code_from_repo(
+            log.warning("Not configured for use with private GitHub repos")
     except Exception as e:
-        msg = f'Unable to setup SSH for Github and you are trying to connect via SSH: {e}' # noqa
+        msg = f"Unable to setup SSH for Github and you are trying to connect via SSH: {e}"  # noqa
```
Probably don't need the `# noqa` after it has been Blacked?

Marios85 replied:
👍

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +50,162 @@ def download_project_code_from_repo(
+            known_hosts.touch(0o700, exist_ok=False)
+            known_hosts.write_text(get_ssh_public_key_from_domain(hostname))
+        elif not known_hosts_contains_domain_key(hostname, known_hosts):
+            known_hosts.touch(0o700)
```
Why do you touch it again? You know it exists if you end-up here.

Marios85 replied:
That's a misunderstanding of the touch command on my part (it doesn't exist on Windows) I thought the mode was also looking at the file access rights but it doesn't.

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +50,162 @@ def download_project_code_from_repo(
+
+
+def get_ssh_public_key_from_domain(hostname: str) -> str:
+    """Gets the public SSH key from the host and checks if the fingerprint matches
```
Need to format the docstring (same for the above). I'll see if I can find a linter.

Marios85 replied:
Have added a space between the first line and the description otherwise I don't what the docstring requirements are.

AlexIoannides commented on .circleci/config.yml:
```diff
@@ -28,36 +28,50 @@ jobs:
       - checkout
+      - add_ssh_keys:
+          fingerprints:
+            - "3b:3f:26:dc:81:3b:6d:49:88:8e:da:f0:d5:9f:99:d6"
```
Which key does this correspond to?

The one with the fingerprint beginning `e2` is the GitLab one, but I don't recognise the one above.

Marios85 replied:
It's the Github key (look at SSH keys in CircleCI project settings). 

However I think it is better to change this to the Github Deploy key if it has access to the private repo as well.

Marios85 commented:
Refactored tests to include a global conftest and also expanded the test teardown to prevent tests affecting each other.

Fixed a Windows OS only bug in Workflow Execution when cleaning up the cloned repo.

AlexIoannides commented on .circleci/config.yml:
```diff
@@ -25,32 +25,45 @@ jobs:
       - run: 
           name: Running tests
-          command: tox -e py38-unit_and_functional_tests,py38-static_code_analysis -- -s -vv
+          command: tox -e py38-unit_and_functional_tests,py38-static_code_analysis -- -s -vv --junitxml=~/tox/junit.xml
```
Why do you need to use `--junitxml` to create test results and why are you switching to the output path afterwards?

Marios85 replied:
For CircleCI to generate the test summary (the tests tab in the build).

AlexIoannides commented on src/bodywork/constants.py:
```diff
@@ -23,19 +23,29 @@
+PROJECT_CONFIG_FILENAME = "bodywork.yaml"
+SSH_DIR_NAME = ".ssh_bodywork"
+SSH_PRIVATE_KEY_ENV_VAR = "BODYWORK_GIT_SSH_PRIVATE_KEY"
+SSH_SECRET_NAME = "ssh-github-private-key"
```
Should be `SSH_SECRET_NAME = "ssh-git-private-key"`.

Or something like that.

Marios85 replied:
A rogue github reference!

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +51,166 @@ def download_project_code_from_repo(
-            log.warning('Not configured for use with private GitHub repos')
+        if get_connection_protocol(url) is ConnectionProtocol.SSH:
+            hostname = urlparse(f"ssh://{url}").hostname
+            if hostname is not None and hostname != "":
```
You could abbreviate this line with `if hostname:`.

You were probably not aware, that `None` is cast to `False`, likewise for `""` or any empty list/container.

The section on `truthiness` in [this post](https://blog.jrheard.com/truthiness-and-short-circuit-evaluation-in-python) is worth reading, but the whole article is worth some time at some point.

Marios85 replied:
Noted and corrected!

AlexIoannides commented on src/bodywork/git.py:
```diff
@@ -44,102 +51,166 @@ def download_project_code_from_repo(
     and then tells Git to use it for SSH by exporting the
     GIT_SSH_COMMAND environment variable.
+
+    :param hostname:    Hostname to SSH to.
```
Leading spaces before `Hostname`

AlexIoannides commented on src/bodywork/workflow_execution.py:
```diff
@@ -273,15 +303,31 @@ def _print_logs_to_stdout(namespace: str, job_or_deployment_name: str) -> None:
+def remove_readonly(func, path, exc_info):
+    """Error handler for ``shutil.rmtree``.
+
+    If the error is due to an access error (read only file) it attempts to add write
```
line length should be 72 (sorry...).

AlexIoannides commented on tests/conftest.py:
```diff
@@ -0,0 +1,94 @@
+import shutil
```
I think we need the licensing blurb up-top here.

AlexIoannides commented on tests/conftest.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@fixture(scope="function")
+def test_namespace() -> str:
```
Do we use this anywhere?

Marios85 replied:
Yes, it's used heavily in test_k8s_cluster.py.

AlexIoannides commented on tests/integration/test_git_integration.py:
```diff
@@ -0,0 +1,92 @@
+"""
```
License text blurb required.

AlexIoannides commented on tests/integration/test_git_integration.py:
```diff
@@ -0,0 +1,92 @@
+"""
+Integration Tests for interactions with hosted Git repositories.
```
`tests` not `Tests`.

Marios85 replied:
Eagle eye 👁️

AlexIoannides commented on tests/integration/test_git_integration.py:
```diff
@@ -0,0 +1,92 @@
+            shutil.rmtree(ssh_dir, onerror=remove_readonly)
+
+
+def remove_readonly(func, path, exc_info):
```
You've got this function in the global `contest.py`. You should be able to import this, if we add `__init__.py` files to the test directories (turning them into proper Python modules and sub-modules, etc.).

Marios85 replied:
This function shouldn't have to be in the global conftest.py but it's there because there are integration tests in the unit tests. The real solution is to move those tests out into Integration and/or make them proper unit tests which should be done as part of that ticket or as and when we touch them as part of normal development. 

I suggest leaving this change until then, when it will be removed.

AlexIoannides reviewed (approved):
Good to go once a few minor changes are made.
